scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,366 @@
+use crate::cdc_types::{GenerationTimestamp, StreamID};
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoint {
+    pub timestamp: Duration,
+    pub stream_id: StreamID,
+    pub generation: GenerationTimestamp,
+}
+
+/// Spawns a new task that periodically saves checkpoints.
+/// Takes vector of tracked by a [`StreamReader`] stream IDs.
+/// Receives value of the last checkpoint via channel.
+/// We are using [`tokio::sync::watch`] channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_saving_checkpoints(
+    tracked_streams: Vec<StreamID>,
+    checkpoint_saver: Arc<dyn CDCCheckpointSaver>,
+    receiver: tokio::sync::watch::Receiver<Checkpoint>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            for stream in tracked_streams.iter() {
+                let mut checkpoint = receiver.borrow().clone();
+                checkpoint.stream_id = stream.clone();
+
+                checkpoint_saver.save_checkpoint(&checkpoint).await?;
```

Comment:
If `save_checkpoint` returns an error, the tokio task will return an error and progress will no longer be saved for this set of streams. Moreover, the error is returned in the handle, but the handle is discarded and error is not checked.

Please change this function so that it doesn't stop after encountering an error - instead, log it with `warn!` level (use the `tracing` crate which was introduced by another PR).